### PR TITLE
fix: zero value nonce is invalid

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Description

Because `0` is the zero (default) value for `uint64`, if it is valid to be used as a nonce, it becomes difficult to distinguish the scenario where sender did not set a nonce from one where they explicitly set it to `0`.

I'm not confident whether the ability to make this distinction matters now or has the potential to later but was following a feeling.

### Summary generated by Reviewpad on 11 Jul 23 11:56 UTC
This pull request includes the following changes:

- In the `p2p/module.go` file, the `handlePocketEnvelope` function now checks if `poktEnvelope.Nonce` is zero and returns an error if it is. The error message includes the hex-encoded nonce value.
- In the `p2p/module_test.go` file, new test cases have been added to test the handling of an invalid nonce.
- In the `p2p/types/errors.go` file, a new error variable `ErrInvalidNonce` has been added to represent an invalid nonce value.
- In the `shared/crypto/rand.go` file, a check has been added to ensure that the generated nonce value is not zero. If it is zero, the function recursively calls itself to generate a new nonce.

## Issue

N/A; observation made while working on #732.

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Adds `ErrInvalidNonce` P2P error type
- Ensures the P2P message handler rejects the zero value `Nonce` (`uint64(0)`) is invalid
- Ensures the `GetNonce` function never returns the zero value

## Testing

- [ ] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/module.go [hunk 1]
```diff
@@ -374,6 +374,14 @@ func (m *p2pModule) handlePocketEnvelope(pocketEnvelopeBz []byte) error {
 		return fmt.Errorf("decoding network message: %w", err)
 	}
 
+	if poktEnvelope.Nonce == 0 {
```
Comment (dylanlott) [inline p2p/module.go:377]: 🤝 🙌

File: p2p/module_test.go
Comment (dylanlott) [inline p2p/module_test.go:1]: Love a change with this test to code ratio 😎

File: p2p/module_test.go [hunk 2]
```diff
@@ -176,46 +248,30 @@ func TestP2pModule_WithHostOption_Restart(t *testing.T) {
...
 
 // TECHDEBT(#609): move & de-duplicate
-func newLibp2pMockNetHost(t *testing.T, privKey cryptoPocket.PrivateKey, peer *typesP2P.NetworkPeer) libp2pHost.Host {
+func newMockNetHost(
+	t *testing.T,
+	libp2pMockNet mocknet.Mocknet,
+	privKey cryptoPocket.PrivateKey,
+	peer *typesP2P.NetworkPeer,
+) libp2pHost.Host {
+	t.Helper()
```
Comment (dylanlott) [inline p2p/module_test.go:267]: Nice 💯

File: p2p/module_test.go
Comment (Olshansk) [reply p2p/module_test.go:1]: There are only two types of changes in this world:
1. Lots of code w/ no tests
2. One line change w/ lots of tests

File: p2p/module_test.go [hunk 3]
```diff
@@ -151,9 +159,73 @@ func Test_Create_configureBootstrapNodes(t *testing.T) {
...
+		t, privKey,
+		WithHostOption(host),
+	)
+	err := mod.Start()
+	require.NoError(t, err)
+
+	poktEnvelope := &messaging.PocketEnvelope{
+		Content: &anypb.Any{},
+		// 0 should be an invalid nonce value
+		Nonce: 0,
```
Comment (Olshansk) [inline p2p/module_test.go:214]: Can you please add another few lines in this test to make sure the default (when we don't set the nonce at all), it is also invalid?

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Review by Olshansk — APPROVED** [review]

**Pull request merged**